File: UVa/524/524.cpp
#include <bits/stdc++.h>

using namespace std;

vector <vector <int>> sols;
int n;

bool esPrimo(int num){
	if(num == 2 || num == 3 || num == 5 || num == 7 || num == 11 || num == 13
		|| num == 17 || num == 19 || num == 23 || num == 29 || num == 31 || num == 37)
		return true;
	return false;
}

void solve(vector <int> visitados, vector <int> sol){
	int tam = sol.size();
	cout << tam << endl;

	if(tam < n){
		for(int i = 2; i <= n; i++){
			if(!visitados[i] && esPrimo(i + sol[tam - 1])){
				visitados[i] = 1;
				sol.push_back(i);
				solve(visitados, sol);
				sol.pop_back();
				visitados[i] = 0;
			}
		}
	}
	else{
		if(esPrimo(sol[0] + sol[tam - 1]))
			sols.push_back(sol);
	}
}

int main(){
	int casos = 1;
	while(cin >> n){
		if(casos >= 2)
			printf("\n");
		vector <int> visitados;
		
		for(int i = 0; i <= n + 4; i++)
			visitados.push_back(0);

		solve(visitados, vector <int> (1));

		cout << "Case " << casos++ << ":" << '\n';
		for(int i = 0; i < sols.size(); i++){
			for(int j = 0; j < n; j++){
				if(j == 0)
					cout << sols[i][j];
				else
					cout << " " << sols[i][j];
			}
			cout << endl;
		}

		sols = vector <vector<int>> ();
	}

	return 0;
}

File: UVa/524/524.py
from sys import stdin
num = 0
sols = []

def esPrimo(num):
	if(num == 2 or num == 3 or num == 5 or num == 7 or num == 11 or num == 13
		or num == 17 or num == 19 or num == 23 or num == 29 or num == 31 or num == 37):
		return True
	return False

def solve(sol, visitados):
	global num, sols
	if(len(sol) < num):
		for i in range(2, num + 1):
			if(visitados[i] == 0 and esPrimo(i + sol[-1])):
				scopy = sol.copy()
				scopy.append(i)
				vcopy = visitados.copy()
				vcopy[i] = 1
				solve(scopy, vcopy)
	else:
		if(esPrimo(sol[0] + sol[-1])):
			sols.append(sol)

def main():
	global num, visitados, sols
	num = stdin.readline()
	cases = 1

	while(len(num) > 0):
		if(cases != 1):
			print()
		visitados = [0 for i in range(18)]
		num = int(num)
		arr = [1]
		solve(arr, visitados)
		print("Case " + str(cases) + ":")
		for i in range(len(sols)):
			for j in range(len(sols[i])):
				if(j != 0):
					print(" " + str(sols[i][j]), end = "")
				else:
					print(sols[i][j], end = "")
			print()

		num = stdin.readline()
		sols = []
		cases += 1
	

main()
